### PR TITLE
Done three-auth task

### PR DESCRIPTION
**Pull request by mnbeh**

### CODE REVIEW

File: packages/social-trader/src/utils/helpers.ts [hunk 1]
```diff
@@ -179,6 +179,19 @@ export const getRandomWords = (length: number = getRandomInteger(3, 8)) =>
 export const getRandomEmail = () =>
   `${getRandomWord()}@${getRandomWord()}.${getRandomWord()}`;
 
+export const partiallyHideEmail = (email: string): string => {
+  if (!email) {
+    return "";
+  }
+  const [first, second] = email.split("@");
+  return `${first.slice(0, -3)}***@***${second.slice(3)}`;
+  const starsAmountBefore: number = Math.floor(first.length / 3);
```
Comment (pavelzubov) [inline packages/social-trader/src/utils/helpers.ts:188]: It doesn't work.

File: packages/social-trader/src/utils/helpers.ts [hunk 2]
```diff
@@ -179,6 +179,19 @@ export const getRandomWords = (length: number = getRandomInteger(3, 8)) =>
 export const getRandomEmail = () =>
   `${getRandomWord()}@${getRandomWord()}.${getRandomWord()}`;
 
+export const partiallyHideEmail = (email: string): string => {
+  if (!email) {
```
Comment (pavelzubov) [inline packages/social-trader/src/utils/helpers.ts:183]: `email` is a required argument. You don't need this check.

File: packages/social-trader/src/pages/auth/security-verification/security-verification-container.tsx [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        successMessage: "auth:security-verification.success-alert-message",
+        request: values => {
+            return confirmThreeStepAuth(values)
+                .then(res => {
+                    if (res) {
+                        return res;
+                    }
+                    Push(LOGIN_ROUTE);
+                })
+                .catch((e: ResponseError) => setErrorMessage(e));
```
Comment (pavelzubov) [inline packages/social-trader/src/pages/auth/security-verification/security-verification-container.tsx:37]: useApiRequest has its own errorMessage.

File: packages/social-trader/src/pages/auth/security-verification/security-verification-container.tsx [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    const { sendRequest, data } = useApiRequest({
+        middleware: [updateTokenMiddleware],
+        successMessage: "auth:security-verification.success-alert-message",
+        request: values => {
+            return confirmThreeStepAuth(values)
+                .then(res => {
+                    if (res) {
+                        return res;
+                    }
+                    Push(LOGIN_ROUTE);
```
Comment (pavelzubov) [inline packages/social-trader/src/pages/auth/security-verification/security-verification-container.tsx:35]: It must be in middleware

File: packages/social-trader/src/pages/auth/security-verification/service/security-verification.service.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    .auth()
+    .confirmThreeStepAuth({
+      body: {
+        token,
+        email,
+        code
+      }
+    })
+    .then(response => {
+      authService.storeToken(response);
```
Comment (pavelzubov) [inline packages/social-trader/src/pages/auth/security-verification/service/security-verification.service.ts:23]: It must be in middleware

File: packages/social-trader/src/pages/auth/signin/signin.service.ts [hunk 6]
```diff
@@ -76,6 +77,31 @@ export const useTwoFactorState = (ctx?: NextPageContext) => {
   };
 };
 
+export const initialThreeFactorState = {
+  email: "test@email.com",
```
Comment (pavelzubov) [inline packages/social-trader/src/pages/auth/signin/signin.service.ts:81]: Why is the initial value of the email not empty?

File: packages/social-trader/src/utils/helpers.ts [hunk 7]
```diff
@@ -179,6 +179,15 @@ export const getRandomWords = (length: number = getRandomInteger(3, 8)) =>
 export const getRandomEmail = () =>
   `${getRandomWord()}@${getRandomWord()}.${getRandomWord()}`;
 
+export const partiallyHideEmail = (email: string): string => {
+  const [first, second] = email.split("@");
+  const starsAmountBefore: number = Math.floor(first.length / 3);
+  const starsAmountAfter: number = Math.floor(second.length / 3);
+  return `${first.slice(0, first.length - starsAmountBefore)}${"*".repeat(
```
Comment (pavelzubov) [inline packages/social-trader/src/utils/helpers.ts:186]: This mask doesn't match the requirement of the task.

File: packages/social-trader/src/utils/helpers.ts [hunk 8]
```diff
@@ -179,6 +179,10 @@ export const getRandomWords = (length: number = getRandomInteger(3, 8)) =>
 export const getRandomEmail = () =>
   `${getRandomWord()}@${getRandomWord()}.${getRandomWord()}`;
 
+export const partiallyHideEmail = (email: string): string => {
+  return email.replace(/(?<=^[A-Za-z0-9]{3}).*?(?=@)/, "***");
```
Comment (pavelzubov) [inline packages/social-trader/src/utils/helpers.ts:183]: image

File: packages/social-trader/src/utils/helpers.ts [hunk 9]
```diff
@@ -179,6 +179,11 @@ export const getRandomWords = (length: number = getRandomInteger(3, 8)) =>
 export const getRandomEmail = () =>
   `${getRandomWord()}@${getRandomWord()}.${getRandomWord()}`;
 
+export const partiallyHideEmail = (email: string): string => {
+  const [first, second] = email.split("@");
+  return `${first.slice(0, 3)}${"*".repeat(3)}@${second}`;
```
Comment (pavelzubov) [inline packages/social-trader/src/utils/helpers.ts:184]: partiallyHideEmail("bob@mail.ru") -> bob***@mail.ru

How long will I be test a framework for you?)))
Please, write tests 🙏🙏🙏

File: packages/social-trader/public/static/locales/en/auth.json [hunk 10]
```diff
@@ -108,6 +108,14 @@
       "continue": "Continue"
     }
   },
+  "security-verification": {
+    "title": "Security verification",
+    "success-alert-message": "Email address successfully verified",
+    "message": "Enter the 6 digit code received by email",
```
[the remark this answers is not in the document]
Comment (mnbeh) [reply packages/social-trader/public/static/locales/en/auth.json:114]: The task description has been changed. Email doesn't need to be hidden anymore.

[the remark this answers is not in the document]
Comment (pavelzubov) [reply packages/social-trader/public/static/locales/en/auth.json:114]: A good solution to problems with partiallyHideEmail function 😂😂😂
image

**Pull request merged**